File: Programmers/문제풀이_구/Level 2/Level2_택배상자.py
# 스택인데?
'''
보조 컨테이너 벨트는 앞 뒤로 이동이 가능하지만 입구 외에 다른 면이 막혀 있어서 
맨 앞의 상자만 뺄 수 있습니다 (즉, 가장 마지막에 보조 컨테이너 벨트에 보관한 상자부터 꺼내게 됩니다).
'''

order = [4, 3, 1, 2, 5]
order = [5, 4, 3, 2, 1]

# 1트(성공), 문제에서 풀이 방법을 쉽게 알려줌
def solution(order):
    answer = 0
    stack = []
    for i in range(len(order)):
        if order[answer] == (i + 1):
            answer += 1
        else:
            stack.append(i + 1)
            continue

        while stack:
            if order[answer] == stack[-1]:
                stack.pop()
                answer += 1
            else:
                break

    return answer

print("RESULT : " + str(solution(order)))

File: Programmers/Level 2/Level2_택배상자.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <queue>
#include <stack>
#include <map>
#include <set>
using namespace std;

int solution(vector<int> order) 
{
    int answer = 0;

    // 보조 컨테이너
	stack<int> station;
    
    for (int i = 1; i <= order.size(); i++) 
    {
		// 현재 상자가 트럭에 실어야 하는 순서
        if (i == order[answer])
			answer++;
		else
			station.push(i);

		// 보조 컨테이너에 있는 상자들 중에서 트럭에 실어야 하는 순서가 있다면
        while (!station.empty() && station.top() == order[answer]) 
        {
            station.pop();
            answer++;
		}

        if (order[answer] < i)
            break;
	}

    return answer;
}

// 다른 사람의 깔끔한 풀이
int other_solution(vector<int> order) 
{
    int answer = 0;
    stack<int> s;

    for (int i = 1; i <= order.size(); i++) 
    {
        s.push(i);

        while (!s.empty() && s.top() == order[answer]) 
        {
            s.pop();
            answer++;
        }
    }

    return answer;
}

// 1차 시도 : 성공, 문제 자체에 대놓고 스택 사용하라고 적혀 있었지만
// 너무 쓸데 없이 설명이 들어가 있어서 이해가 오히려 안됐던 문제